File: cli/internal/projectconfig/projectconfig.go
package projectconfig

import (
	"bufio"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// ProjectConfig represents the per-project .cixconfig.yaml file.
type ProjectConfig struct {
	Ignore IgnoreConfig `yaml:"ignore"`
}

// IgnoreConfig controls which files/directories are excluded from indexing.
type IgnoreConfig struct {
	Submodules bool `yaml:"submodules"`
}

// Load reads .cixconfig.yaml from the project root.
// Returns a zero-value config (not an error) if the file does not exist.
func Load(projectRoot string) (ProjectConfig, error) {
	var cfg ProjectConfig

	data, err := os.ReadFile(filepath.Join(projectRoot, ".cixconfig.yaml"))
	if err != nil {
		if os.IsNotExist(err) {
			return cfg, nil
		}
		return cfg, err
	}

	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return cfg, err
	}

	return cfg, nil
}

// SubmodulePaths parses .gitmodules and returns the list of submodule paths.
// Returns nil (not an error) if .gitmodules does not exist.
func SubmodulePaths(projectRoot string) ([]string, error) {
	f, err := os.Open(filepath.Join(projectRoot, ".gitmodules"))
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	defer f.Close()

	var paths []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if strings.HasPrefix(line, "path") {
			parts := strings.SplitN(line, "=", 2)
			if len(parts) == 2 {
				p := strings.TrimSpace(parts[1])
				if p != "" {
					paths = append(paths, p)
				}
			}
		}
	}

	return paths, scanner.Err()
}

File: legacy/python-api/app-root/app/services/project_config.py
import re
from dataclasses import dataclass, field
from pathlib import Path

import yaml


@dataclass
class IgnoreConfig:
    submodules: bool = False


@dataclass
class ProjectConfig:
    ignore: IgnoreConfig = field(default_factory=IgnoreConfig)


def load_project_config(project_root: str) -> ProjectConfig:
    """Load .cixconfig.yaml from the project root.
    Returns default config if the file does not exist."""
    config_path = Path(project_root) / ".cixconfig.yaml"
    if not config_path.exists():
        return ProjectConfig()

    try:
        data = yaml.safe_load(config_path.read_text())
    except Exception:
        return ProjectConfig()

    if not isinstance(data, dict):
        return ProjectConfig()

    ignore_data = data.get("ignore", {})
    return ProjectConfig(
        ignore=IgnoreConfig(
            submodules=bool(ignore_data.get("submodules", False)),
        )
    )


def parse_submodule_paths(project_root: str) -> list[str]:
    """Parse .gitmodules and return list of submodule paths.
    Returns empty list if .gitmodules does not exist."""
    gitmodules_path = Path(project_root) / ".gitmodules"
    if not gitmodules_path.exists():
        return []

    paths: list[str] = []
    try:
        for line in gitmodules_path.read_text().splitlines():
            line = line.strip()
            if line.startswith("path"):
                parts = line.split("=", 1)
                if len(parts) == 2:
                    p = parts[1].strip()
                    if p:
                        paths.append(p)
    except Exception:
        pass

    return paths